gax/src/main/java/com/google/api/gax/retrying/RetryAlgorithm.java

```diff
@@ -116,13 +183,48 @@ public boolean shouldRetry(
         && timedAlgorithm.shouldRetry(nextAttemptSettings);
   }
 
+  /**
+   * Returns {@code true} if another attempt should be made, or {@code false} otherwise.
+   *
+   * @param context the {@link RetryingContext} that can be used to determine whether another
+   *     attempt should be made
+   * @param previousThrowable exception thrown by the previous attempt or null if a result was
+   *     returned instead
+   * @param previousResponse response returned by the previous attempt or null if an exception was
+   *     thrown instead
+   * @param nextAttemptSettings attempt settings, which will be used for the next attempt, if
+   *     accepted
+   * @throws CancellationException if the retrying process should be cancelled
+   * @return {@code true} if another attempt should be made, or {@code false} otherwise
+   */
+  public boolean shouldRetry(
+      RetryingContext context,
+      Throwable previousThrowable,
+      ResponseT previousResponse,
+      TimedAttemptSettings nextAttemptSettings)
+      throws CancellationException {
+    return getResultAlgorithmWithContext().shouldRetry(context, previousThrowable, previousResponse)
+        && nextAttemptSettings != null
+        && getTimedAlgorithmWithContext().shouldRetry(context, nextAttemptSettings);
+  }
+
   @BetaApi("Surface for inspecting the a RetryAlgorithm is not yet stable")
   public ResultRetryAlgorithm<ResponseT> getResultAlgorithm() {
     return resultAlgorithm;
   }
 
+  @BetaApi("Surface for inspecting the a RetryAlgorithm is not yet stable")
```

Comment:
Added new method instead of changing the return type of the existing one to avoid a binary breaking change.